syntax = "proto3";

package msgpb2;

// Task-related messages
message Task {
    enum TaskType {
        TASK_TYPE_FULL = 0;
        TASK_TYPE_DELTA = 1;
    }

    string message_id = 1;
    int64 message_time = 2;
    string task_id = 3;
    TaskType task_type = 4;
    repeated SourceDetail sources = 5;
    int64 deadline = 6;
    bool cancelled = 7;
}

message SourceDetail {
    enum SourceType {
        SOURCE_TYPE_REPO = 0;
        SOURCE_TYPE_FUZZ_TOOLING = 1;
        SOURCE_TYPE_DIFF = 2;
    }

    string sha256 = 1;
    SourceType source_type = 2;
    string url = 3;
}

message TaskDownload {
    Task task = 1;
}

message TaskReady {
    Task task = 1;
}

message TaskDelete {
    string task_id = 1;
    float received_at = 2;
}

// Fuzzer-related messages
message BuildRequest {
    string package_name = 1;
    string engine = 2;
    string sanitizer = 3;
    string ossfuzz = 4;
    string source_path = 5;
    string task_id = 6;
    string build_type = 7;
}

message BuildOutput {
    string package_name = 1;
    string engine = 2;
    string sanitizer = 3;
    string output_ossfuzz_path = 4;
    string source_path = 5;
    string task_id = 6;
    string build_type = 7;
}

message WeightedHarness {
    float weight = 1;
    string package_name = 2;
    string harness_name = 3;
    string task_id = 4;
}

message Crash {
    BuildOutput target = 1;
    string harness_name = 2;
    string crash_input_path = 3;
    string stacktrace = 4;
} 

message ConfirmedVulnerability {
    Crash crash = 1;
    string vuln_id = 2;
}

message Patch {
    string task_id = 1;
    string vulnerability_id = 2;
    string patch = 3;
}

message IndexRequest {
    string package_name = 1;
    string ossfuzz = 2;
    string source_path = 3;
    string task_id = 4;
    string build_type = 5;
}
